otlp-logger.js

```diff
@@ -0,0 +1,153 @@
+ */
+function getOtlpLogger (opts) {
+  const loggerProvider = new LoggerProvider({
+    resource: new Resource({
+      'service.name': opts.serviceName,
+      'service.version': opts.serviceVersion
+    })
+  })
+
+  const recordProcessor =
+    opts.useBatchProcessor ?? true
+      ? new BatchLogRecordProcessor(opts.logRecordExporter)
+      : new SimpleLogRecordProcessor(opts.logRecordExporter)
+
+  loggerProvider.addLogRecordProcessor(recordProcessor)
+
+  logs.setGlobalLoggerProvider(loggerProvider)
+
+  const logger = logs.getLogger(opts.loggerName, opts.serviceVersion, {
+    includeTraceContext: opts.includeTraceContext ?? true
+  })
+
+  const mapperOptions = {
+    messageKey: opts.messageKey || DEFAULT_MESSAGE_KEY
+  }
+
+  return {
+    /**
+     * @param {Bindings} obj
+     */
+    emit (obj) {
+      logger.emit(toOpenTelemetry(obj, mapperOptions))
+    },
+    async shutdown () {
+      return loggerProvider.shutdown()
+    }
+  }
+}
+
+/**
+ * If the source format has only a single severity that matches the meaning of the range
+ * then it is recommended to assign that severity the smallest value of the range.
+ * https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#mapping-of-severitynumber
+ */
+const SEVERITY_NUMBER_MAP = {
+  10: SeverityNumber.TRACE,
+  20: SeverityNumber.DEBUG,
+  30: SeverityNumber.INFO,
+  40: SeverityNumber.WARN,
+  50: SeverityNumber.ERROR,
+  60: SeverityNumber.FATAL
+}
+
+// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#displaying-severity
+const SEVERITY_NAME_MAP = {
+  1: 'TRACE',
+  2: 'TRACE2',
+  3: 'TRACE3',
+  4: 'TRACE4',
+  5: 'DEBUG',
+  6: 'DEBUG2',
+  7: 'DEBUG3',
+  8: 'DEBUG4',
+  9: 'INFO',
+  10: 'INFO2',
+  11: 'INFO3',
+  12: 'INFO4',
+  13: 'WARN',
+  14: 'WARN2',
+  15: 'WARN3',
+  16: 'WARN4',
+  17: 'ERROR',
+  18: 'ERROR2',
+  19: 'ERROR3',
+  20: 'ERROR4',
+  21: 'FATAL',
+  22: 'FATAL2',
+  23: 'FATAL3',
+  24: 'FATAL4'
+}
+
+/**
+ * @typedef {Object} CommonBindings
+ * @property {string=} msg
+ * @property {number=} level
+ * @property {number=} time
+ * @property {string=} hostname
+ * @property {number=} pid
+ *
+ * @typedef {Record<string, string | number | Object> & CommonBindings} Bindings
+ *
+ */
+
+/**
+ * Converts a pino log object to an OpenTelemetry log object.
+ *
+ * @typedef {Object} MapperOptions
+ * @property {string} messageKey
+ *
+ * @param {Bindings} sourceObject
+ * @param {MapperOptions} mapperOptions
+ * @returns {import('@opentelemetry/api-logs').LogRecord}
+ */
+function toOpenTelemetry (sourceObject, { messageKey }) {
+  const {
+    time,
+    level,
+    hostname,
+    pid,
+    [messageKey]: msg,
+    ...attributes
+  } = sourceObject
+
+  const severityNumber = SEVERITY_NUMBER_MAP[sourceObject.level] ?? SEVERITY_NUMBER_MAP[60]
+  const severityText = SEVERITY_NAME_MAP[severityNumber]
+
+  return {
+    timestamp: time * 10e6,
```

Comment:
Is this needed? OTel will add a timestamp

Reply:
In the resulting logs, There are two timestamps:

1. timeUnixNano - one that is provided by the user and specified in the data model as [Timestamp](https://github.com/open-telemetry/opentelemetry-specification/blob/fc8289b8879f3a37e1eba5b4e445c94e74b20359/specification/logs/data-model.md#field-timestamp)
2. observedTimeUnixNano - one that is added by OTLP SDK and specified in the data model as [ObservedTimestamp](https://github.com/open-telemetry/opentelemetry-specification/blob/fc8289b8879f3a37e1eba5b4e445c94e74b20359/specification/logs/data-model.md#field-observedtimestamp)


in 1. we are sending the time as observed by Pino.

Reply:
The SDK adds a timestamp in the unix nano format. The observed timestamp is added by the collector, not the SDK

Reply:
Can you clarify what the problem is with having both fields?h The [docs for the field](https://github.com/open-telemetry/opentelemetry-specification/blob/fc8289b8879f3a37e1eba5b4e445c94e74b20359/specification/logs/data-model.md#field-timestamp) say:
_This field is optional, it may be missing if the source timestamp is unknown._

in our case, the source timestamp is known, so why not set it? Each has its use and means different things. If you run the example from the Readme, you will get something like the following logRecords (in `/tmp/test-logs/otlp-logs.log`)



```
          "logRecords": [
            {
              "timeUnixNano": "1690326690783000064",
              "observedTimeUnixNano": "1690326690964999936",
...
            },
            {
              "timeUnixNano": "1690326691784000000",
              "observedTimeUnixNano": "1690326691996000000",
...
            },
            {
              "timeUnixNano": "1690326692784000000",
              "observedTimeUnixNano": "1690326693028999936",
...
            },
            {
              "timeUnixNano": "1690326693784999936",
              "observedTimeUnixNano": "1690326694059000064",
...
            },
            {
              "timeUnixNano": "1690326694785999872",
              "observedTimeUnixNano": "1690326695091000064",
...
            }


```
I assume one signifies the pino logger's time, while the other signifies the time when the batch containing the log record was sent.

This is how the values get set initially:

https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/sdk-logs/src/LogRecord.ts#L88

So the behaviour you recommend is a fallback behaviour. I'd rather go with capturing more info in the log.

Reply:
Thanks for the suggestions, If I missed something here, please create an Issue, and we can continue there. I'll merge the PR so that we do not expand the scope of this PR indefinitely